### PR TITLE
Fix premature creation of file read streams

### PR DESCRIPTION
**Pull request by christensson**
Read stream now created on the finish event for the
write stream to the temp file. The finish event is
emitted once the file is completely flushed.

Also, the promise returned from the exported function
isn't resolved until all files have been downloaded.

### CODE REVIEW

**Comment by christensson** [comment]
Missed PR #14 which does the same thing...

Anyhow, please look at this PR and merge and/or suggest changes. The master branch contains bugs. Try running the tests in this PR a couple of times without the suggested changes in index.js and you'll see what I mean.

File: index.js [hunk 1]
```diff
@@ -88,21 +92,26 @@ function onField(fields, name, val, fieldnameTruncated, valTruncated) {
   }
 }
 
-function onFile(files, fieldname, file, filename, encoding, mimetype) {
+function onFile(filePromises, fieldname, file, filename, encoding, mimetype) {
   const tmpName = file.tmpName = new Date().getTime()  + fieldname  + filename;
   const saveTo = path.join(os.tmpdir(), path.basename(tmpName));
-  file.on('end', () => {
-    const readStream = fs.createReadStream(saveTo);
```
Comment (christensson) [inline index.js]: Possible that readStream is created before `saveTo` is completely written.

**Comment by christensson** [comment]
Fixes issue #7

**Comment by dominhhai** [comment]
Same here https://github.com/m4nuC/async-busboy/pull/14

**Comment by m4nuC** [comment]
Good catch, thanks guys.

File: index.js [hunk 2]
```diff
@@ -56,8 +56,12 @@ module.exports = function (request, options) {
 
     function onEnd(err) {
       if(err) reject(err);
-      cleanup();
-      return resolve({fields, files});
+      Promise.all(filePromises)
```
Comment (dominhhai) [inline index.js]: Empty `filePromises` should `resolve` immediately

Comment (christensson) [inline index.js]: Exactly and it does that as well. Any problems?

The following code prints "Promise result []"
```js
Promise.resolve([]).then(console.log.bind(null,"Promise result"))
```

Comment (dominhhai) [inline index.js]: Performance should be considered here, IMO.

Comment (christensson) [inline index.js]: That is a big sub-optimization. If you want performance you should rewrite to get rid of the temporary file stored in /tmp.

Comment (dominhhai) [inline index.js]: Using `Promise` with empty `filePromises` is slower than `resolve` immediately.
Smt like:
```
if (filePromises.length > 0) {
  return Promise.all(filePromises)
                        .then(files => resolve({fields, files}))
                        .catch(reject)
 } else {
  return resolve({fields, files: []})
}
```

Comment (m4nuC) [inline index.js]: @dominhhai, good point, but I am wondering if that is worth adding branching in the code.  Let's put a pin on that, if perf ever becomes a problem, we can pull this back. Thanks for the insight!

**Pull request merged**